framework/stats/vulkan_stats_provider.cpp

```diff
@@ -0,0 +1,460 @@
+	{
+		stat_data.clear();
+		counter_indices.clear();
+		return;
+	}
+
+	// These stats are fully supported by this provider and in a single pass, so remove
+	// from the requested set.
+	// Subsequent providers will then only look for things that aren't already supported.
+	for (const auto &s : stat_data)
+		requested_stats.erase(s.first);
+}
+
+VulkanStatsProvider::~VulkanStatsProvider()
+{
+	if (stat_data.size() > 0)
+	{
+		// Release profiling lock
+		vkReleaseProfilingLockKHR(device.get_handle());
+	}
+}
+
+bool VulkanStatsProvider::fill_vendor_data()
+{
+	const auto &pd_props = device.get_gpu().get_properties();
+	if (pd_props.vendorID == 0x14E4)        // Broadcom devices
+	{
+		LOGI("Using Vulkan performance counters from Broadcom device");
+
+		// NOTE: The names here are actually regular-expressions.
+		// Counter names can change between hardware variants for the same vendor,
+		// so regular expression names mean that multiple h/w variants can be easily supported.
+		// clang-format off
+		vendor_data = {
+		    {StatIndex::gpu_cycles,          {"cycle_count"}},
+		    {StatIndex::gpu_vertex_cycles,   {"qpu_active_cycles_vertex_coord_user"}},
+		    {StatIndex::gpu_fragment_cycles, {"qpu_active_cycles_fragment"}},
+		    {StatIndex::gpu_fragment_jobs,   {"render_jobs_completed"}},
+		    {StatIndex::gpu_ext_reads,       {"gpu_mem_reads"}},
+		    {StatIndex::gpu_ext_writes,      {"gpu_mem_writes"}},
+		    {StatIndex::gpu_ext_read_bytes,  {"gpu_bytes_read"}},
+		    {StatIndex::gpu_ext_write_bytes, {"gpu_bytes_written"}},
+		};
+		// clang-format on
+
+		// Override vendor-specific graph data
+		vendor_data.at(StatIndex::gpu_vertex_cycles).set_vendor_graph_data({"Vertex/Coord/User Cycles", "{:4.0f}/s"});
+		vendor_data.at(StatIndex::gpu_fragment_jobs).set_vendor_graph_data({"Render Jobs", "{:4.0f}/s"});
+
+		return true;
+	}
+#if 0
+	else if (pd_props.vendorID == xxxx) // Other vendor's devices
+	{
+		// Fill vendor_data for other vendor
+		return true;
+	}
+#endif
+	return false;        // Unsupported vendor
+}
+
+bool VulkanStatsProvider::create_query_pools(uint32_t num_framebuffers, uint32_t queue_family_index)
+{
+	const PhysicalDevice &gpu = device.get_gpu();
+
+	// Now we know the available counters, we can build a query pool that will collect them.
+	// We will check that the counters can be collected in a single pass. Multi-pass would
+	// be a big performance hit so for these samples, we don't want to use it.
+	VkQueryPoolPerformanceCreateInfoKHR perf_create_info{};
+	perf_create_info.sType             = VK_STRUCTURE_TYPE_QUERY_POOL_PERFORMANCE_CREATE_INFO_KHR;
+	perf_create_info.queueFamilyIndex  = queue_family_index;
+	perf_create_info.counterIndexCount = uint32_t(counter_indices.size());
+	perf_create_info.pCounterIndices   = counter_indices.data();
+
+	uint32_t passes_needed;
+	vkGetPhysicalDeviceQueueFamilyPerformanceQueryPassesKHR(gpu.get_handle(), &perf_create_info,
```

Comment:
Could this function be part of the physical device class?

Reply:
Done